### PR TITLE
Step 3

### PR DESCRIPTION
**Pull request by tonklable**
## What

- Build environment for Go
- add GET /items
- edit POST /items to read and write file
- add GET /items/:itemId

## CHECKS :warning:

Please make sure you are aware of the following.

- [ ] **The changes in this PR doesn't have private information

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (karuna) [inline .vscode/settings.json]: If this file is personal settings, it's better to not commit and ignore this file. Git is used to synchronize between people.

File: go/app/main.go [hunk 2]
```diff
@@ -51,6 +134,36 @@ func getImg(c echo.Context) error {
...
+
+func saveItemToJSON(jsonItems Items) {
+	// Save data to JSON file
+	data, _ := json.Marshal(jsonItems)
+	_ = os.WriteFile("items.json", data, 0644)
+}
+
+func calculateImageHash(imageFilePath string) string {
+	//Read image file
+	imageData, _ := os.ReadFile(imageFilePath)
```
Comment (karuna) [inline go/app/main.go]: it's possible an error occurred here, and error should be handled. `calculateImageHash` is used on `addItem`, and `addItem` returns error. Why not handle and propagate the error to `addItem` ?

File: go/app/main.go [hunk 3]
```diff
@@ -27,13 +42,81 @@ func root(c echo.Context) error {
...
 func addItem(c echo.Context) error {
 	// Get form data
-	name := c.FormValue("name")
-	c.Logger().Infof("Receive item: %s", name)
+	var newItem ItemDetail
+	newItem.Name = c.FormValue("name")
+	newItem.Category = c.FormValue("category")
+	imgFilePath := c.FormValue("image")
+	hash := calculateImageHash(imgFilePath)
+	newItem.ImageFilename = hash
```
Comment (karuna) [inline go/app/main.go]:
```
hash, err := calculateImageHash(imgFilePath)
// Handle err here
newItem := {
  Name: c.FormValue("name"),
  Category: c.FormValue("category"),
  ImageFilename: hash,
}
```
The first one is we're handling the error, for example when the storage is full and we cannot create more image.
The second one is we're composing the `newItem` at the same time, so it's clearer.

File: go/app/main.go [hunk 4]
```diff
@@ -51,6 +134,36 @@ func getImg(c echo.Context) error {
...
+	data, _ := os.ReadFile("items.json")
+
+	//Parse JSON data
+	var jsonItems Items
+	_ = json.Unmarshal(data, &jsonItems)
+
+	return jsonItems
+}
+
+func saveItemToJSON(jsonItems Items) {
```
Comment (karuna) [inline go/app/main.go]: same, it's better to handle error here.

File: go/app/main.go [hunk 5]
```diff
@@ -27,13 +42,81 @@ func root(c echo.Context) error {
...
+	newItem.ImageFilename = hash
+
+	// Add new item to existing items
+	existingItems := loadItemsFromJSON()
+	existingItems.Items = append(existingItems.Items, newItem)
+
+	saveItemToJSON(existingItems)
+
+	c.Logger().Infof("Receive item: %s", newItem)
+	// message := fmt.Sprintf("Item %s added", newItem.Name)
```
Comment (karuna) [inline go/app/main.go]: what is this for? unused comment should not be committed.

File: go/app/main.go [hunk 6]
```diff
@@ -27,13 +42,81 @@ func root(c echo.Context) error {
...
+	json.Unmarshal(jsonData, &jsonContent)
+	return c.JSON(http.StatusOK, jsonContent)
+}
+
+func getItemDetail(c echo.Context) error {
+	// Get item ID
+	itemId := c.Param("itemId")
+
+	// Change string to int
+	itemIdInt, _ := strconv.Atoi(itemId)
```
Comment (karuna) [inline go/app/main.go]: same here, error handling

**Comment by tonklable** [comment]
@karuna Could you please check again?

**Pull request merged**